File: pushTemplates/src/main/java/com/clevertap/android/pushtemplates/AsyncHelper.java
package com.clevertap.android.pushtemplates;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

class AsyncHelper {

    private long EXECUTOR_THREAD_ID = 0;
    private final ExecutorService es;

    private static AsyncHelper asyncHelperInstance = null;

    private AsyncHelper() {
        this.es = Executors.newFixedThreadPool(1);
    }

    static AsyncHelper getInstance() {
        if (asyncHelperInstance == null) {
            asyncHelperInstance = new AsyncHelper();
        }
        return asyncHelperInstance;
    }

    @SuppressWarnings({"UnusedParameters", "SameParameterValue"})
    void postAsyncSafely(final String name, final Runnable runnable) {
        try {
            final boolean executeSync = Thread.currentThread().getId() == EXECUTOR_THREAD_ID;

            if (executeSync) {
                runnable.run();
            } else {
                this.es.submit(new Runnable() {
                    @Override
                    public void run() {
                        EXECUTOR_THREAD_ID = Thread.currentThread().getId();
                        try {
                            runnable.run();
                        } catch (Throwable t) {
                            PTLog.verbose("Executor service: Failed to complete the scheduled task" + name);
                        }
                    }
                });
            }
        } catch (Throwable t) {
            PTLog.verbose("Failed to submit task to the executor service");
        }
    }

    public static Handler getMainThreadHandler() {
        Handler mainThreadHandler;
        mainThreadHandler = new Handler(Looper.getMainLooper());
        return mainThreadHandler;
    }
}

File: pushTemplates/src/main/java/com/clevertap/android/pushtemplates/AsyncHelper.kt
package com.clevertap.android.pushtemplates

import android.os.Handler
import android.os.Looper
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

internal class AsyncHelper private constructor() {
    private var EXECUTOR_THREAD_ID: Long = 0
    private val es: ExecutorService
    fun postAsyncSafely(name: String, runnable: Runnable) {
        try {
            val executeSync = Thread.currentThread().id == EXECUTOR_THREAD_ID
            if (executeSync) {
                runnable.run()
            } else {
                es.submit {
                    EXECUTOR_THREAD_ID = Thread.currentThread().id
                    try {
                        runnable.run()
                    } catch (t: Throwable) {
                        PTLog.verbose("Executor service: Failed to complete the scheduled task$name")
                    }
                }
            }
        } catch (t: Throwable) {
            PTLog.verbose("Failed to submit task to the executor service")
        }
    }

    companion object {
        private var asyncHelperInstance: AsyncHelper? = null
        @JvmStatic
        val instance: AsyncHelper?
            get() {
                if (asyncHelperInstance == null) {
                    asyncHelperInstance = AsyncHelper()
                }
                return asyncHelperInstance
            }
        @JvmStatic
        val mainThreadHandler: Handler
            get() {
                val mainThreadHandler: Handler
                mainThreadHandler = Handler(Looper.getMainLooper())
                return mainThreadHandler
            }
    }

    init {
        es = Executors.newFixedThreadPool(1)
    }
}